### PR TITLE
✨ Implement numerical and linalg features

### PR DESCRIPTION
**Pull request by Bal7hazar**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: #123

## What is the new behavior?

- Linear algebra

  - Dot product

- Numerical analysis

  - Trapezoidal rule
  - Cumsum a sequence
  - Diff a sequence
  - Interpolate y at x with different interpolation and extrapolation methods

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Developments done with:
```bash
scarb 0.5.0-alpha.4 (6484cf7b6 2023-06-29)
cairo: 2.0.0-rc6 (https://crates.io/crates/cairo-lang-compiler/2.0.0-rc6)
```

Workload: ~ 6 hours

### CODE REVIEW

**Comment by Bal7hazar** [comment]
@LucasLvy I'm done for a first review, I create 2 new folders (`numeric` and `linalg`) because they respectively have high growth potential, not sure if you prefered to have it in math.

File: alexandria/numeric/src/interp.cairo [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            Extrapolation::Constant(()) => *ys.at(xs.len() - 1_usize),
+        };
+        return y;
+    }
+
+    // [Compute] Interpolation
+    let mut index = 0_usize;
+    loop {
+        if index + 1_usize == xs.len() {
+            assert(0 != 1, 'Unexpected error');
```
Comment (LucasLvy) [inline alexandria/numeric/src/interp.cairo:58]:
```suggestion
            assert(0 == 1, 'Unexpected error');
```
You mean that right ?

File: alexandria/numeric/src/interp.cairo [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        }
+        assert(*xs.at(index + 1_usize) > *xs.at(index), 'Abscissa must be sorted');
```
Comment (LucasLvy) [inline alexandria/numeric/src/interp.cairo:60]: you should be able to use that:
```suggestion
        assert(*xs[index + 1_usize] > *xs.at(index), 'Abscissa must be sorted');
```

**Review by LucasLvy — CHANGES_REQUESTED** [review]
Looks good but switch to rust doc (`///` instead of `//` above functions). And please use more explicit function names

File: alexandria/numeric/src/interp.cairo (hunk 2 above)
Comment (Bal7hazar) [reply alexandria/numeric/src/interp.cairo:60]: Thank you, I'll do that

File: alexandria/numeric/src/interp.cairo (hunk 1 above)
Comment (Bal7hazar) [reply alexandria/numeric/src/interp.cairo]: I replaced it by `false`  the I just removed it since this case cannot be reached logically.

**Comment by Bal7hazar** [comment]
> Looks good but switch to rust doc (/// instead of // above functions). And please use more explicit function names
I replaced `interp`  by `interpolate` and `trapz`  to `trapezoidal_rule` which break from numpy naming, but it's ok I think.

Also `dot`, `cumsum`  and `diff` sounds to be explicit enough to me (it is named like this in other scientific libs I think). But I can change them also if it matters.

**Review by LucasLvy — APPROVED** [review]

**Pull request merged**